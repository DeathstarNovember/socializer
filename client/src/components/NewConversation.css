.new-conversation {
  border: 1px solid rgba(0, 0, 0, 0.125);
}
