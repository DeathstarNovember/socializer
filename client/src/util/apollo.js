import ApolloClient from "apollo-client";
import { InMemoryCache } from "apollo-cache-inmemory";
import * as AbsintheSocket from "@absinthe/socket";
import { createAbsintheSocketLink } from "@absinthe/socket-apollo-link";
import { Socket as PhoenixSocket, LongPoll } from "phoenix";
import { createHttpLink } from "apollo-link-http";
import { hasSubscription } from "@jumpn/utils-graphql";
import { split } from "apollo-link";
import { setContext } from "apollo-link-context";
import Cookies from "js-cookie";

const HTTP_URI =
  process.env.NODE_ENV === "production"
    ? "https://brisk-hospitable-indianelephant.gigalixirapp.com"
    : "http://localhost:4000";

const WS_URI =
  process.env.NODE_ENV === "production"
    ? "wss://brisk-hospitable-indianelephant.gigalixirapp.com/socket"
    : "ws://localhost:4000/socket";

function createPhoenixSocket() {
  // Socket with full fallback to LongPoll
  // via https://elixirforum.com/t/fall-back-to-longpoll-when-websocket-fails/23894
  const socket = new PhoenixSocket(WS_URI, {
    params: () => {
      if (Cookies.get("token")) {
        return { token: Cookies.get("token") };
      } else {
        return {};
      }
    },
  });
  socket.onError(() => {
    if (navigator.onLine) {
      const isWebSocket = socket.transport === window.WebSocket;
      console.error(
        `Error connecting using ${isWebSocket ? "WebSocket" : "LongPoll"}`,
      );
      if (isWebSocket) socket.transport = LongPoll;
      else if (window.WebSocket) socket.transport = window.WebSocket;
    }
  });
  return socket;
}

export const createClient = ({ ssr, req, fetch, tokenCookie } = {}) => {
  let link = createHttpLink({ uri: HTTP_URI, fetch });
  let absintheSocket;

  if (!ssr) {
    absintheSocket = AbsintheSocket.create(createPhoenixSocket());
    const socketLink = createAbsintheSocketLink(absintheSocket);

    link = split(
      (operation) => hasSubscription(operation.query),
      socketLink,
      link,
    );
  }

  const authLink = setContext((_, { headers }) => {
    // Get the authentication token from the cookie if it exists.
    const token = tokenCookie || Cookies.get("token");

    // Return the headers to the context so httpLink can read them.
    return {
      headers: {
        ...headers,
        authorization: token ? `Bearer ${token}` : "",
      },
    };
  });

  const cache = new InMemoryCache();
  if (!ssr) {
    cache.restore(window.__APOLLO_STATE__);
  }

  const client = new ApolloClient({
    connectToDevTools: !ssr,
    ssrMode: ssr,
    link: authLink.concat(link),
    cache,
  });
  return { client, absintheSocket };
};

export const refreshSocket = (socket) => {
  // Close the connection to force a reconnection with the
  // new token parameter.
  socket.phoenixSocket.conn && socket.phoenixSocket.conn.close();
};
