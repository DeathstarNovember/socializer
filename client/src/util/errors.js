import { camelizeKeys } from "humps";

export const errorHash = ({ graphQLErrors } = {}) => {
  if (!Array.isArray(graphQLErrors)) return {};
  return camelizeKeys(
    graphQLErrors.reduce(
      (hash, error) => Object.assign(hash, { [error.field]: error.message }),
      {},
    ),
  );
};
