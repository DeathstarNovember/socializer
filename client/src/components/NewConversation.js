import React, { useContext, useState } from "react";
import { useQuery, useMutation } from "@apollo/react-hooks";
import gql from "graphql-tag";
import { Button } from "react-bootstrap";
import { Redirect } from "react-router-dom";
import Select from "react-select";
import renderIf from "render-if";
import { ErrorMessage } from "components";
import { ChatContext } from "util/context";
import "./NewConversation.css";

export const SEARCH_USERS = gql`
  query SearchUsers($searchTerm: String!) {
    searchUsers(searchTerm: $searchTerm) {
      id
      name
      gravatarMd5
    }
  }
`;

export const CREATE_CONVERSATION = gql`
  mutation CreateConversation($userIds: [String]!) {
    createConversation(userIds: $userIds) {
      id
    }
  }
`;

const NewConversation = () => {
  const { setChatState } = useContext(ChatContext);
  const [searchTerm, setSearchTerm] = useState("");
  const [users, setUsers] = useState([]);
  const {
    loading: searchLoading,
    error: searchError,
    data: searchData,
  } = useQuery(SEARCH_USERS, {
    variables: { searchTerm },
  });
  const [create, { data, loading }] = useMutation(CREATE_CONVERSATION);
  if (data) {
    setChatState("default");
    return <Redirect to={`/chat/${data.createConversation.id}`} />;
  }

  return (
    <div className="new-conversation p-2">
      {renderIf(searchError)(() => (
        <ErrorMessage message={searchError.message} />
      ))}
      {renderIf(!searchError)(
        <Select
          isLoading={searchLoading}
          isMulti
          inputValue={searchTerm}
          placeholder="Select users..."
          value={users}
          onChange={(value) => setUsers(value)}
          onInputChange={(value) => setSearchTerm(value)}
          options={
            searchData.searchUsers &&
            searchData.searchUsers.map((user) => ({
              label: user.name,
              value: user.id,
            }))
          }
        />,
      )}
      <div className="d-flex justify-content-between mt-2">
        <Button variant="danger" onClick={() => setChatState("default")}>
          Cancel
        </Button>

        <Button
          variant="primary"
          disabled={loading}
          onClick={() =>
            create({
              variables: { userIds: users.map((user) => user.value) },
            })
          }
        >
          {loading ? "Creating..." : "Create"}
        </Button>
      </div>
    </div>
  );
};

export default NewConversation;
