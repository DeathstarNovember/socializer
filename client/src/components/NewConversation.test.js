import React from "react";
import { render, fireEvent, wait } from "@testing-library/react";
import { MockedProvider } from "@apollo/react-testing";
import { MemoryRouter } from "react-router-dom";
import { ChatContext } from "util/context";
import NewConversation, {
  SEARCH_USERS,
  CREATE_CONVERSATION,
} from "./NewConversation";

jest.mock("react-router-dom", () => ({
  MemoryRouter: jest.requireActual("react-router-dom").MemoryRouter,
  Redirect: ({ to }) => <div data-testid="redirect" data-to={to} />,
}));

describe("NewConversation", () => {
  it("renders correctly", () => {
    const mocks = [
      {
        request: {
          query: SEARCH_USERS,
          variables: { searchTerm: "" },
        },
        result: { data: { searchUsers: [] } },
      },
    ];
    const { container } = render(
      <ChatContext.Provider value={{ setChatState: jest.fn() }}>
        <MockedProvider mocks={mocks} addTypename={false}>
          <NewConversation />
        </MockedProvider>
      </ChatContext.Provider>,
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it("switches back to list when button is clicked", () => {
    const mocks = [
      {
        request: {
          query: SEARCH_USERS,
          variables: { searchTerm: "" },
        },
        result: { data: { searchUsers: [] } },
      },
    ];
    const setChatState = jest.fn();
    const { getByText } = render(
      <ChatContext.Provider value={{ setChatState }}>
        <MockedProvider mocks={mocks} addTypename={false}>
          <NewConversation />
        </MockedProvider>
      </ChatContext.Provider>,
    );
    fireEvent.click(getByText("Cancel"));
    expect(setChatState).toBeCalledWith("default");
  });

  it("creates new conversation", async () => {
    const mocks = [
      {
        request: {
          query: SEARCH_USERS,
          variables: { searchTerm: "" },
        },
        result: { data: { searchUsers: [] } },
      },
      {
        request: {
          query: SEARCH_USERS,
          variables: { searchTerm: "Jeremy" },
        },
        result: {
          data: {
            searchUsers: [{ id: 1, name: "Jeremy Lane", gravatarMd5: "abc" }],
          },
        },
      },
      {
        request: {
          query: SEARCH_USERS,
          variables: { searchTerm: "" },
        },
        result: { data: { searchUsers: [] } },
      },
      {
        request: {
          query: SEARCH_USERS,
          variables: { searchTerm: "John" },
        },
        result: {
          data: {
            searchUsers: [{ id: 2, name: "John Thomas", gravatarMd5: "def" }],
          },
        },
      },
      {
        request: {
          query: SEARCH_USERS,
          variables: { searchTerm: "" },
        },
        result: { data: { searchUsers: [] } },
      },
      {
        request: {
          query: CREATE_CONVERSATION,
          variables: { userIds: [1, 2] },
        },
        result: { data: { createConversation: { id: 123 } } },
      },
    ];
    const { container, getByTestId, getByText } = render(
      <MemoryRouter>
        <ChatContext.Provider value={{ setChatState: jest.fn() }}>
          <MockedProvider mocks={mocks} addTypename={false}>
            <NewConversation />
          </MockedProvider>
        </ChatContext.Provider>
      </MemoryRouter>,
    );

    const input = container.querySelector("input");
    fireEvent.change(input, { target: { value: "Jeremy" } });
    await wait(() => fireEvent.click(getByText("Jeremy Lane")));
    fireEvent.change(input, { target: { value: "John" } });
    await wait(() => fireEvent.click(getByText("John Thomas")));
    expect(container.firstChild).toMatchSnapshot();
    fireEvent.click(getByText("Create"));
    await wait(() =>
      expect(getByTestId("redirect").dataset.to).toEqual("/chat/123"),
    );
  });
});
