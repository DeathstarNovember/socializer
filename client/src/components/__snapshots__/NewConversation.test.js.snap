// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewConversation creates new conversation 1`] = `
<div
  class="new-conversation p-2"
>
  <div
    class="css-10nd86i"
  >
    <div
      class="css-vj8t7z"
    >
      <div
        class="css-1hwfws3"
      >
        <div
          class="css-xwjg1b"
        >
          <div
            class="css-12jo7m5"
          >
            Jeremy Lane
          </div>
          <div
            class="css-1alnv5e"
          >
            <svg
              aria-hidden="true"
              class="css-19bqh2r"
              focusable="false"
              height="14"
              viewBox="0 0 20 20"
              width="14"
            >
              <path
                d="M14.348 14.849c-0.469 0.469-1.229 0.469-1.697 0l-2.651-3.030-2.651 3.029c-0.469 0.469-1.229 0.469-1.697 0-0.469-0.469-0.469-1.229 0-1.697l2.758-3.15-2.759-3.152c-0.469-0.469-0.469-1.228 0-1.697s1.228-0.469 1.697 0l2.652 3.031 2.651-3.031c0.469-0.469 1.228-0.469 1.697 0s0.469 1.229 0 1.697l-2.758 3.152 2.758 3.15c0.469 0.469 0.469 1.229 0 1.698z"
              />
            </svg>
          </div>
        </div>
        <div
          class="css-xwjg1b"
        >
          <div
            class="css-12jo7m5"
          >
            John Thomas
          </div>
          <div
            class="css-1alnv5e"
          >
            <svg
              aria-hidden="true"
              class="css-19bqh2r"
              focusable="false"
              height="14"
              viewBox="0 0 20 20"
              width="14"
            >
              <path
                d="M14.348 14.849c-0.469 0.469-1.229 0.469-1.697 0l-2.651-3.030-2.651 3.029c-0.469 0.469-1.229 0.469-1.697 0-0.469-0.469-0.469-1.229 0-1.697l2.758-3.15-2.759-3.152c-0.469-0.469-0.469-1.228 0-1.697s1.228-0.469 1.697 0l2.652 3.031 2.651-3.031c0.469-0.469 1.228-0.469 1.697 0s0.469 1.229 0 1.697l-2.758 3.152 2.758 3.15c0.469 0.469 0.469 1.229 0 1.698z"
              />
            </svg>
          </div>
        </div>
        <div
          class="css-1g6gooi"
        >
          <div
            class=""
            style="display: inline-block;"
          >
            <input
              aria-autocomplete="list"
              autocapitalize="none"
              autocomplete="off"
              autocorrect="off"
              id="react-select-4-input"
              spellcheck="false"
              style="box-sizing: content-box; width: 2px; border: 0px; font-size: inherit; opacity: 1; outline: 0; padding: 0px;"
              tabindex="0"
              type="text"
              value=""
            />
            <div
              style="position: absolute; top: 0px; left: 0px; visibility: hidden; height: 0px; overflow: scroll; white-space: pre; font-size: inherit; font-family: -webkit-small-control; letter-spacing: normal; text-transform: none;"
            />
          </div>
        </div>
      </div>
      <div
        class="css-1wy0on6"
      >
        <div
          aria-hidden="true"
          class="css-fraxkc"
        >
          <span
            class="css-pujnn1"
          />
          <span
            class="css-gji98i"
          />
          <span
            class="css-1frpy8x"
          />
        </div>
        <span
          class="css-d8oujb"
        />
        <div
          aria-hidden="true"
          class="css-1ep9fjw"
        >
          <svg
            aria-hidden="true"
            class="css-19bqh2r"
            focusable="false"
            height="20"
            viewBox="0 0 20 20"
            width="20"
          >
            <path
              d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
            />
          </svg>
        </div>
      </div>
    </div>
  </div>
  <div
    class="d-flex justify-content-between mt-2"
  >
    <button
      class="btn btn-danger"
      type="button"
    >
      Cancel
    </button>
    <button
      class="btn btn-primary"
      type="button"
    >
      Create
    </button>
  </div>
</div>
`;

exports[`NewConversation renders correctly 1`] = `
<div
  class="new-conversation p-2"
>
  <div
    class="css-10nd86i"
  >
    <div
      class="css-vj8t7z"
    >
      <div
        class="css-1hwfws3"
      >
        <div
          class="css-1492t68"
        >
          Select users...
        </div>
        <div
          class="css-1g6gooi"
        >
          <div
            class=""
            style="display: inline-block;"
          >
            <input
              aria-autocomplete="list"
              autocapitalize="none"
              autocomplete="off"
              autocorrect="off"
              id="react-select-2-input"
              spellcheck="false"
              style="box-sizing: content-box; width: 2px; border: 0px; font-size: inherit; opacity: 1; outline: 0; padding: 0px;"
              tabindex="0"
              type="text"
              value=""
            />
            <div
              style="position: absolute; top: 0px; left: 0px; visibility: hidden; height: 0px; overflow: scroll; white-space: pre; font-size: inherit; font-family: -webkit-small-control; letter-spacing: normal; text-transform: none;"
            />
          </div>
        </div>
      </div>
      <div
        class="css-1wy0on6"
      >
        <div
          aria-hidden="true"
          class="css-fraxkc"
        >
          <span
            class="css-pujnn1"
          />
          <span
            class="css-gji98i"
          />
          <span
            class="css-1frpy8x"
          />
        </div>
        <span
          class="css-d8oujb"
        />
        <div
          aria-hidden="true"
          class="css-1ep9fjw"
        >
          <svg
            aria-hidden="true"
            class="css-19bqh2r"
            focusable="false"
            height="20"
            viewBox="0 0 20 20"
            width="20"
          >
            <path
              d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
            />
          </svg>
        </div>
      </div>
    </div>
  </div>
  <div
    class="d-flex justify-content-between mt-2"
  >
    <button
      class="btn btn-danger"
      type="button"
    >
      Cancel
    </button>
    <button
      class="btn btn-primary"
      type="button"
    >
      Create
    </button>
  </div>
</div>
`;
